src/frontend/src/hooks/UseGenerator.tsx

```diff
@@ -0,0 +1,70 @@
+import { useQuery } from '@tanstack/react-query';
+import { useCallback, useState } from 'react';
+
+import { api } from '../App';
+import { ApiEndpoints } from '../enums/ApiEndpoints';
+import { apiUrl } from '../states/ApiState';
+
+export type GeneratorState = {
+  query: Record<string, any>;
+  result: any;
+  update: (params: Record<string, any>, overwrite?: boolean) => void;
+};
+
+/* Hook for managing generation of data via the InvenTree API.
+ * We pass an endpoint, and start with an initially empty query.
+ * We can pass additional parameters to the query, and update the query as needed.
+ * Each update calls a new query to the API, and the result is stored in the state.
+ */
+export function useGenerator(
+  endpoint: ApiEndpoints,
+  key: string,
+  onGenerate?: (value: any) => void
+): GeneratorState {
+  // Track the result
+  const [result, setResult] = useState<any>(null);
+
+  // Track the generator query
+  const [query, setQuery] = useState<Record<string, any>>({});
+
+  // Callback to update the generator query
+  const update = useCallback(
+    (params: Record<string, any>, overwrite?: boolean) => {
+      if (overwrite ?? false) {
+        setQuery(params);
+      } else {
+        setQuery({
+          ...query,
+          ...params
+        });
+      }
+
+      queryGenerator.refetch();
+    },
+    [query]
+  );
```

Comment:
You could make the ref to the update function static if you would do something like this. This would have the advantage, that the useMemo from the form fields doesn't need to trigger all the time the query state updates.
```suggestion
  const update = useCallback(
    (params: Record<string, any>, overwrite?: boolean) => {
      if (overwrite) {
        setQuery(params);
      } else {
        setQuery((query) => ({
          ...query,
          ...params
        }));
      }

      queryGenerator.refetch();
    },
    []
  );
```